src/sql/rewrite/ob_transform_simplify_expr.cpp

```diff
@@ -2517,4 +2526,681 @@ int ObTransformSimplifyExpr::try_remove_ora_decode(ObRawExpr *&expr,
+ * @discription: according to the results of when exprs
+ *  1. cond1 ... cond(k-1) is FALSE/NULL
+ *  2. condk is TRUE/UNCALCULABLE
+ *    2.1 if condk is TRUE, then rewrite case_expr to exprk
+ *    2.2 if condk is UNCALCULABLE, then remove cond1...cond(k-1)
+ *    2.3 if k > when_cnt, then rewrite case_expr to default_expr
+ */
+int ObTransformSimplifyExpr::convert_case_when_predicate_by_when_exprs(
+                                                      ObRawExpr *&expr, 
+                                                      bool &trans_happened) {
+  int ret = OB_SUCCESS;
+
+  trans_happened = false;
+  ObCaseOpRawExpr *case_expr = NULL;
+
+  bool has_null = false; // check cond1 to cond(k-1) has null expr or not
+  int64_t first_true_non_calc_idx = -1; // record the index of first true or non_calc expr
+  ObSEArray<ObRawExpr*, 2> false_null_exprs; // used to add constraints
+
+  if (OB_ISNULL(expr) || OB_ISNULL(ctx_)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("param is NULL", K(ret));
+  } else if (T_OP_CASE != expr->get_expr_type()) {
+  } else if (OB_FALSE_IT(case_expr = static_cast<ObCaseOpRawExpr *>(expr))) {
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() != case_expr->get_then_expr_size())) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("incorrect case when params", K(ret));
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() < 1)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("case when size is less than 1", K(ret));
+  } else {
+    bool is_true = false;   // check condk is true expr or not
+    bool is_uncalc = false; // check condk is uncalculable or not
+    bool is_false_or_null = false;  // check condk is false or null expr 
+    int64_t idx = 0; // first non false idx
+    int64_t when_cnt = case_expr->get_when_expr_size();
+    ObSEArray<ObRawExpr*, 2> false_exprs; // used to add constraints. nvl(expr, false)
+
+    for (idx = 0; OB_SUCC(ret) && idx < when_cnt; ++idx) {
+      if (OB_FAIL(check_when_exprs_validity(
+                                        case_expr->get_when_param_expr(idx),
+                                        is_true,
+                                        is_uncalc,
+                                        is_false_or_null))) {
+        LOG_WARN("failed to check when expr validity");
+      } else if (OB_LIKELY(is_false_or_null)) {
+        ObRawExpr *nvl_false_expr = NULL;
+        if (OB_FAIL(build_nvl_expr(case_expr->get_when_param_expr(idx), 
+                                   nvl_false_expr, 
+                                   false))) {
+          LOG_WARN("failed to build nvl expr",K(ret));
+        } else if (OB_FAIL(false_exprs.push_back(nvl_false_expr))) {
+          LOG_WARN("failed to push back false_exprs", K(ret));
+        }
+      } else {
+        break;
+      }
+    }
+
+    if (OB_FAIL(ret)) {
+    } else if (OB_UNLIKELY(idx < when_cnt && !is_true && !is_uncalc)) {
+    } else if (OB_FAIL(add_false_or_false_constraint(false_exprs))) {
```

Comment:
add and use add_constraint_for_convert_case_when to add constraint for convert_case_when_predicate_by_when_exprs and convert_case_when_predicate_by_then_exprs ？